### PR TITLE
Rename `uniform` to `parameter` across the engine

### LINKED ISSUE #5123: Rename API around shaders to use the word "parameter" and not "uniform" (except text shaders)
### Describe the project you are working on

VFX

### Describe the problem or limitation you are having in your project

"Uniform" is a highly specific keyword which isn't really related to its meaning. Godot is the only widespread engine to use it and it's highly counter intuitive.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Prefer the keyword "parameter" over uniform. Specifically:
1. ShadeMaterial `set_shader_uniform` should be named back to `set_shader_param`
2. VisualShader FloatUniform, ColorUniform
3. Global/Instance uniform -> param

Text shaders still use the keyword **uniform**!!!

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See above

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Changes core API

### Is there a reason why this should be core and not an add-on in the asset library?

Changes core API

### PR DESCRIPTION
**Pull request by Chaosus**
Changes:

- `ShaderMaterial::set_shader_uniform/get_shader_uniform` renamed to `set_shader_param/get_shader_param`.
- All other visual shader classes `<Type>Uniform` renamed to `<Type>Parameter`.
- Other uniform methods, parameter and variables renamed to parameters in the `VisualShaderEditor`.
- Separates `VisualShaderNodeTextureUniform` to `VisualShaderNodeTextureParameter` and `VisualShaderNodeTexture2DParameter`.
- Changed `default_value` type in `VisualShaderNodeVec4Parameter` from `Quaternion` to `Vector4`.
- Added compatibility classes and changed docs accordingly.

Target to close https://github.com/godotengine/godot-proposals/issues/5123 hope I did everything correct (as far as I remember the last bikeshedding meeting)

### CODE REVIEW

**Comment by QbieShay** [comment]
Hey @Chaosus, thanks a lot for taking the time to do this!
I have two questions:
1. Is the entry for "global shader unifrom" in project setting also changed?
2. I am not sure if they preferred `set_shader_param` or `set_shader_parameter`? (@akien-mga ?)
The rest looks good to me! (Note I didn't test, i can't this weekend)

**Comment by Chaosus** [comment]
```
Is the entry for "global shader unifrom" in project setting also changed?
```
In the project setting, it's called `Shader Globals` - how do you want to change it?

image

Also, I didn't change the name of the methods in the `RenderingServer` and `MaterialStorage` - I'm not sure that it is a good thing to do, and @reduz would certainly against it.

**Comment by clayjohn** [comment]
> Also, I didn't change the name of the methods in the RenderingServer and MaterialStorage - I'm not sure that it is a good thing to do, and @reduz would certainly against it.

I think anything exposed to users should use "parameter". I don't think it is a big deal to change the naming in RenderingServer and MaterialStorage. After all, reduz originally named them with parameter and it was just changed to "uniform" recently (in https://github.com/godotengine/godot/pull/59840 and https://github.com/godotengine/godot/pull/59844)

**Comment by Chaosus** [comment]
Okay, if this is desired I will make a patch to do it ^^

**Comment by Zireael07** [comment]
> I think anything exposed to users should use "parameter". 

Yep, should be consistent whichever way the cookie crumbles

**Comment by Chaosus** [comment]
@clayjohn Alright, I've renamed uniform term to parameter across all functions and enumeration members:

image

**Comment by nathanfranke** [comment]
If anyone disagrees with the naming in https://github.com/godotengine/godot/pull/64092, let me know. I decided `shader_param/` for compatibility/simplicity, and since the internal properties are not user facing (users should really be using `set_shader_param("foo", bar)`)

**Comment by akien-mga** [comment]
> If anyone disagrees with the naming in #64092, let me know. I decided `shader_param/` for compatibility/simplicity, and since the internal properties are not user facing (users should really be using `set_shader_param("foo", bar)`)

We should fully spell out `parameter` everywhere (it's already `set_shader_parameter` in this PR).

It would be useful to have compatibility code for `shader_param` (before #59844) and `shader_uniform` (before this PR) to ease transition from 3.x and earlier alphas.

**Comment by Chaosus** [comment]
I've added `set_shader_param` -> `set_shader_parameter` to the project converter. Also renamed bunch of parameter names and other internal functions. I leave only some internal classes untouched such as `RD::Uniform`, `UniformSet` etc - do I need to rename them too?

**Comment by clayjohn** [comment]
> I've added `set_shader_param` -> `set_shader_parameter` to the project converter. Also renamed bunch of parameter names and other internal functions. I leave only some internal classes untouched such as `RD::Uniform`, `UniformSet` etc - do I need to rename them too?

No, RD::Uniform and Uniform set etc. should all remain as is. Just the user-facing functions to set a single uniform should be changed. I'll review now!

File: drivers/vulkan/rendering_device_vulkan.h [hunk 1]
```diff
@@ -546,7 +546,7 @@ class RenderingDeviceVulkan : public RenderingDevice {
 	};
 
 	struct UniformInfo {
-		UniformType type = UniformType::UNIFORM_TYPE_MAX;
+		ParameterType type = ParameterType::PARAMETER_TYPE_MAX;
```
Comment (clayjohn) [inline drivers/vulkan/rendering_device_vulkan.h:549]: I don't think we should change the RenderingDevice API. Its better to keep this closer to the actual names. It won't be used by beginners either, i.e. you have to know what a uniform is in order to use the RD API

File: scene/resources/shader.cpp [hunk 2]
```diff
@@ -107,7 +107,7 @@ void Shader::get_shader_uniform_list(List<PropertyInfo> *p_params, bool p_get_gr
 	_update_shader();
 
 	List<PropertyInfo> local;
```
Comment (clayjohn) [inline scene/resources/shader.cpp:109]: This function should be renamed to ``get_shader_parameter_list``

File: scene/resources/shader.cpp [hunk 3]
```diff
@@ -176,8 +176,8 @@ bool Shader::is_text_shader() const {
 	return true;
 }
 
-bool Shader::has_uniform(const StringName &p_param) const {
-	return params_cache.has("shader_uniform/" + p_param);
+bool Shader::has_parameter(const StringName &p_name) const {
+	return params_cache.has("shader_uniform/" + p_name);
```
Comment (clayjohn) [inline scene/resources/shader.cpp:180]: Probably should be ``shader_parameter``, but I wonder where the variables are set in the cache

File: scene/resources/material.cpp [hunk 4]
```diff
@@ -384,15 +384,15 @@ void ShaderMaterial::_shader_changed() {
...
 	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "shader", PROPERTY_HINT_RESOURCE_TYPE, "Shader"), "set_shader", "get_shader");
 }
 
 void ShaderMaterial::get_argument_options(const StringName &p_function, int p_idx, List<String> *r_options) const {
 	String f = p_function.operator String();
-	if ((f == "get_shader_uniform" || f == "set_shader_uniform") && p_idx == 0) {
+	if ((f == "get_shader_parameter" || f == "set_shader_parameter") && p_idx == 0) {
 		if (shader.is_valid()) {
 			List<PropertyInfo> pl;
 			shader->get_shader_uniform_list(&pl);
```
Comment (clayjohn) [inline scene/resources/material.cpp:398]: In this function "shader_uniform" should be "shader_parameter"

Comment (clayjohn) [reply scene/resources/material.cpp:398]: Also backwards compatibility should be added to ``_set()`` and ``_get()``

File: servers/register_server_types.cpp [hunk 5]
```diff
@@ -216,7 +216,7 @@ void register_server_types() {
 	GDREGISTER_CLASS(RDFramebufferPass);
 	GDREGISTER_CLASS(RDSamplerState);
 	GDREGISTER_CLASS(RDVertexAttribute);
-	GDREGISTER_CLASS(RDUniform);
+	GDREGISTER_CLASS(RDParameter);
```
Comment (clayjohn) [inline servers/register_server_types.cpp:219]: Should be reverted

File: servers/rendering/rendering_device.h [hunk 6]
```diff
@@ -42,7 +42,7 @@ class RDSamplerState;
 class RDVertexAttribute;
 class RDShaderSource;
 class RDShaderSPIRV;
-class RDUniforms;
```
Comment (clayjohn) [inline servers/rendering/rendering_device.h:45]: All of this should be reverted. RenderingDevice should continue using "uniform"

**Review by clayjohn — COMMENTED** [review]
Looks mostly good. Great work! I think everything in RenderingDevice should continue to use "uniform". So you'll have to undo those changes unfortunately. I think it is best it maintains the original name as people using it will be more familiar with uniforms as uniform has a specific meaning in the context of graphics APIs that is incompatible with the more general term "parameter". Parameter is fine for the Godot-shader API though, so it is suitable for user-facing code in the editer, scene nodes and RenderingServer

**Comment by Chaosus** [comment]
@clayjohn Done I think - check again

**Comment by reduz** [comment]
The actual problem is that uniform term is unfriendly to VisualShader for artists, everywhere else it should be uniform, which is the right term. "parameter" is not more friendly for programmers, its more confusing because anything can be a parameter.

File: drivers/vulkan/rendering_device_vulkan.h [hunk 7]
```diff
@@ -1120,9 +1120,9 @@ class RenderingDeviceVulkan : public RenderingDevice {
 	virtual RID storage_buffer_create(uint32_t p_size_bytes, const Vector<uint8_t> &p_data = Vector<uint8_t>(), uint32_t p_usage = 0);
 	virtual RID texture_buffer_create(uint32_t p_size_elements, DataFormat p_format, const Vector<uint8_t> &p_data = Vector<uint8_t>());
 
-	virtual RID uniform_set_create(const Vector<Uniform> &p_uniforms, RID p_shader, uint32_t p_shader_set);
-	virtual bool uniform_set_is_valid(RID p_uniform_set);
-	virtual void uniform_set_set_invalidation_callback(RID p_uniform_set, InvalidationCallback p_callback, void *p_userdata);
+	virtual RID uniform_set_create(const Vector<Uniform> &p_parameters, RID p_shader, uint32_t p_shader_set);
```
Comment (reduz) [inline drivers/vulkan/rendering_device_vulkan.h:1123]: RenderingDevice needs to be left alone, these are fine.

File: servers/rendering/renderer_geometry_instance.h [hunk 8]
```diff
@@ -59,7 +59,7 @@ class RenderGeometryInstance {
 	virtual void set_use_dynamic_gi(bool p_enable) = 0;
 	virtual void set_use_lightmap(RID p_lightmap_instance, const Rect2 &p_lightmap_uv_scale, int p_lightmap_slice_index) = 0;
 	virtual void set_lightmap_capture(const Color *p_sh9) = 0;
-	virtual void set_instance_shader_uniforms_offset(int32_t p_offset) = 0;
+	virtual void set_instance_shader_parameters_offset(int32_t p_offset) = 0;
```
Comment (reduz) [inline servers/rendering/renderer_geometry_instance.h:62]: This is OK as uniforms

File: servers/rendering/renderer_geometry_instance.h [hunk 9]
```diff
@@ -140,7 +140,7 @@ class RenderGeometryInstanceBase : public RenderGeometryInstance {
 	virtual void set_transparency(float p_transparency) override;
 	virtual void set_use_baked_light(bool p_enable) override;
 	virtual void set_use_dynamic_gi(bool p_enable) override;
-	virtual void set_instance_shader_uniforms_offset(int32_t p_offset) override;
+	virtual void set_instance_shader_parameters_offset(int32_t p_offset) override;
```
Comment (reduz) [inline servers/rendering/renderer_geometry_instance.h:143]: Also OK as uniforms

File: servers/rendering/renderer_geometry_instance.cpp [hunk 10]
```diff
@@ -117,7 +117,7 @@ void RenderGeometryInstanceBase::set_use_dynamic_gi(bool p_enable) {
 	_mark_dirty();
 }
 
-void RenderGeometryInstanceBase::set_instance_shader_uniforms_offset(int32_t p_offset) {
+void RenderGeometryInstanceBase::set_instance_shader_parameters_offset(int32_t p_offset) {
```
Comment (reduz) [inline servers/rendering/renderer_geometry_instance.cpp:120]: This should be ok as uniforms

File: servers/rendering_server.h [hunk 11]
```diff
@@ -1469,20 +1469,20 @@ class RenderingServer : public Object {
...
-	virtual void global_shader_uniform_set_override(const StringName &p_name, const Variant &p_value) = 0;
+	virtual void global_shader_parameter_set(const StringName &p_name, const Variant &p_value) = 0;
+	virtual void global_shader_parameter_set_override(const StringName &p_name, const Variant &p_value) = 0;
 
-	virtual Variant global_shader_uniform_get(const StringName &p_name) const = 0;
-	virtual GlobalShaderUniformType global_shader_uniform_get_type(const StringName &p_name) const = 0;
+	virtual Variant global_shader_parameter_get(const StringName &p_name) const = 0;
+	virtual GlobalShaderParameterType global_shader_parameter_get_type(const StringName &p_name) const = 0;
 
 	virtual void global_shader_uniforms_load_settings(bool p_load_textures) = 0;
```
Comment (reduz) [inline servers/rendering_server.h:1482]: Missed these

**Review by reduz — CHANGES_REQUESTED** [review]
It also marks vulkan_core.h as changed, it should not be?

**Comment by Chaosus** [comment]
```It also marks vulkan_core.h as changed, it should not be?```

Sorry, due to formatting suddenly, reverted it back

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 12]
```diff
@@ -1736,7 +1736,7 @@ void RasterizerSceneGLES3::render_scene(RID p_render_buffers, const CameraData *
 	// Fill Light lists here
 	//////////
 
-	GLuint global_buffer = GLES3::MaterialStorage::get_singleton()->global_shader_uniforms_get_uniform_buffer();
+	GLuint global_buffer = GLES3::MaterialStorage::get_singleton()->global_shader_parameters_get_parameter_buffer();
```
Comment (reduz) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1739]: it should be:
`global_shader_parameters_get_uniform_buffer`

**Review by reduz — CHANGES_REQUESTED** [review]

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**